core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/kopt/KOptDescriptor.java

```diff
@@ -0,0 +1,504 @@
+            removedEdgeIndexToTourOrder[j] =
+                    (endpointToSuccessorFunction.apply(removedEdges[i]) == removedEdges[i + 1]) ? i : i + 1;
+        }
+
+        IntComparator comparator = (pa, pb) -> pa == pb ? 0
+                : (betweenPredicate.test(removedEdges[removedEdgeIndexToTourOrder[1]], removedEdges[pa], removedEdges[pb]) ? -1
+                        : 1);
+        IntArrays.stableSort(removedEdgeIndexToTourOrder, 2, k + 1, comparator);
+
+        for (int i, j = 2 * k; j >= 2; j -= 2) {
+            removedEdgeIndexToTourOrder[j - 1] = i = removedEdgeIndexToTourOrder[j / 2];
+            removedEdgeIndexToTourOrder[j] = ((i & 1) == 1) ? i + 1 : i - 1;
+        }
+
+        for (int i = 1; i <= 2 * k; i++) {
+            inverseRemovedEdgeIndexToTourOrder[removedEdgeIndexToTourOrder[i]] = i;
+        }
+    }
+
+    public int[] getRemovedEdgeIndexToTourOrder() {
+        return removedEdgeIndexToTourOrder;
+    }
+
+    /**
+     * Calculate the disjoint k-cycles for {@link KOptDescriptor#removedEdgeIndexToTourOrder}. <br />
+     * <br />
+     * Any permutation can be expressed as combination of k-cycles. A k-cycle is a sequence of
+     * unique elements (p_1, p_2, ..., p_k) where
+     * <ul>
+     * <li>p_1 maps to p_2 in the permutation</li>
+     * <li>p_2 maps to p_3 in the permutation</li>
+     * <li>p_(k-1) maps to p_k in the permutation</li>
+     * <li>p_k maps to p_1 in the permutation</li>
+     * <li>In general: p_i maps to p_(i+1) in the permutation</li>
+     * </ul>
+     * For instance, the permutation
+     * <ul>
+     * <li>1 -> 2</li>
+     * <li>2 -> 3</li>
+     * <li>3 -> 1</li>
+     * <li>4 -> 5</li>
+     * <li>5 -> 4</li>
+     * </ul>
+     * can be expressed as `(1, 2, 3)(4, 5)`.
+     *
+     * @return The {@link KOptCycleInfo} corresponding to the permutation described by
+     *         {@link KOptDescriptor#removedEdgeIndexToTourOrder}.
+     */
+    public KOptCycleInfo getCyclesForPermutation() {
+        int cycleCount = 0;
+        int[] indexToCycle = new int[removedEdgeIndexToTourOrder.length];
+        IntSet remaining = IntSets.fromTo(1, removedEdgeIndexToTourOrder.length);
+        while (!remaining.isEmpty()) {
+            int current = remaining.intIterator().nextInt();
+            remaining.remove(current);
+
+            while (true) {
+                indexToCycle[current] = cycleCount;
+                int tourIndex = removedEdgeIndexToTourOrder[current];
+                int nextEndpointTourIndex = addedEdgeToOtherEndpoint[tourIndex];
+                current = inverseRemovedEdgeIndexToTourOrder[nextEndpointTourIndex];
+                if (!remaining.contains(current)) {
+                    break;
+                }
+                remaining.remove(current);
+            }
+            cycleCount++;
+        }
+
+        return new KOptCycleInfo(cycleCount, indexToCycle);
+    }
+
+    public List<Pair<Node_, Node_>> getAddedEdges() {
```

Comment:
This method is only used in `toString()`. Let's hide it from the public API, I'd even suggest moving it out of this class entirely into some util class and making it static. 

It would require adding some arguments, but I'd argue that `KOptDescriptor` is already complex enough even without these methods which aren't strictly necessary for its functioning.